packages/contracts-bedrock/scripts/ChainAssertions.sol

```diff
@@ -0,0 +1,178 @@
+import { ResourceMetering } from "src/L1/ResourceMetering.sol";
+import { DeployConfig } from "scripts/DeployConfig.s.sol";
+import { SystemConfig } from "src/L1/SystemConfig.sol";
+import { Constants } from "src/libraries/Constants.sol";
+import { L1StandardBridge } from "src/L1/L1StandardBridge.sol";
+import { L2OutputOracle } from "src/L1/L2OutputOracle.sol";
+import { ProtocolVersion, ProtocolVersions } from "src/L1/ProtocolVersions.sol";
+import { OptimismPortal } from "src/L1/OptimismPortal.sol";
+import { L1CrossDomainMessenger } from "src/L1/L1CrossDomainMessenger.sol";
+import { OptimismMintableERC20Factory } from "src/universal/OptimismMintableERC20Factory.sol";
+import { L1ERC721Bridge } from "src/L1/L1ERC721Bridge.sol";
+import { Predeploys } from "src/libraries/Predeploys.sol";
+import { Types } from "scripts/Types.sol";
+import { Vm } from "forge-std/Vm.sol";
+
+library ChainAssertions {
+    function postDeployAssertions(
+        address systemConfig,
+        address protocolVersions,
+        DeployConfig cfg,
+        uint256 l2OutputOracleStartingTimestamp,
+        Vm vm
+    )
+        internal
+        view
+    {
+        SystemConfig sys = SystemConfig(systemConfig);
+        Types.ContractSet memory proxies = Types.ContractSet({
+            L1CrossDomainMessenger: sys.l1CrossDomainMessenger(),
+            L1StandardBridge: sys.l1StandardBridge(),
+            L2OutputOracle: sys.l2OutputOracle(),
+            OptimismMintableERC20Factory: sys.optimismMintableERC20Factory(),
+            OptimismPortal: sys.optimismPortal(),
+            SystemConfig: address(sys),
+            L1ERC721Bridge: sys.l1ERC721Bridge(),
+            ProtocolVersions: protocolVersions
+        });
+        postDeployAssertions(proxies, cfg, l2OutputOracleStartingTimestamp, vm);
+    }
+
+    function postDeployAssertions(
+        Types.ContractSet memory prox,
+        DeployConfig cfg,
+        uint256 l2OutputOracleStartingTimestamp,
+        Vm vm
+    )
+        internal
+        view
+    {
+        ResourceMetering.ResourceConfig memory rcfg = SystemConfig(prox.SystemConfig).resourceConfig();
+        ResourceMetering.ResourceConfig memory dflt = Constants.DEFAULT_RESOURCE_CONFIG();
+        require(keccak256(abi.encode(rcfg)) == keccak256(abi.encode(dflt)));
+
+        _postCheckSystemConfig(prox, cfg);
+        _postCheckL1CrossDomainMessenger(prox, vm);
+        _postCheckL1StandardBridge(prox, vm);
+        _postCheckL2OutputOracle(prox, cfg, l2OutputOracleStartingTimestamp);
+        _postCheckOptimismMintableERC20Factory(prox);
+        _postCheckL1ERC721Bridge(prox);
+        _postCheckOptimismPortal(prox, cfg);
+        _postCheckProtocolVersions(prox, cfg);
+    }
+
+    /// @notice Post check hook for the system config
+    function _postCheckSystemConfig(Types.ContractSet memory proxies, DeployConfig cfg) private view {
+        SystemConfig config = SystemConfig(proxies.SystemConfig);
```

Comment:
@tynes This assumes the `SystemConfig` has already been upgraded. Is this a fair assumption to make for all multi-chain upgrades?

Reply:
I think that this is fine for now but we may need to decouple this function into 2 functions in the future depending on how our needs evolve. It seems like we will revert the contract address getters on the system config for the next immediate upgrade

Reply:
Good to know. Agreed on decouple this in that case.

Reply:
Introduced a new SystemConfig interface for the version 0 ABI. Scripts can call a ChainAssertions function to check v0 configs. Alternatively, we could read the `veresion()` from the `SystemConfig` to determine which checks will be appropriate.

Reply:
Doing string comparisons in evm is no fun. We didn't really design the semver stuff to be able to work on chain, but we could have a library that parses the semver on chain into native solidity types

Reply:
I think not doing on chain branching logic based on version is fine for now, I think we should write a library that can do protocol versions style version parsing and use that in the long term but for now we can skip it